app/models/filter.rb

```diff
@@ -180,4 +180,46 @@ def allowed_taxonomies
       errors.add(:location_ids, _('You can\'t assign locations to this resource'))
     end
   end
+
+  def fix_to_current_user_permissions
+    return unless role_id
+    return if permission_ids.empty?
+    user = User.current
+    return unless user
+    return if user.admin?
+    current_filters = user.filters.includes(:permissions)
+    current_permissions = Hash[current_filters.map { |f| [f.permissions.first.resource_type, f ]}]
+    new_permissions = permissions
+    new_resource_type = resource_type
+
+    unless current_permissions.has_key? new_resource_type
+      errors.add(:resource_type, (_('Current user is not allowed to resource type: %s') % new_resource_type))
+      return
+    end
+
+    # allow adding filters to permitted resources only
+    allowed_permissions =  new_permissions & current_permissions[new_resource_type].permissions
+    self.permission_ids = allowed_permissions.map(&:id)
+
+    # restrict filter condition if the user has custom filter
+    new_search = search || ''
+    # compare the filters using scoped search tokenizer, to avoid whitespace/syntax differences
+    new_search_tokens = ScopedSearch::QueryLanguage::Compiler.tokenize new_search
+    if current_permissions[new_resource_type].search
+      self.unlimited = 0
+      current_search_tokens = ScopedSearch::QueryLanguage::Compiler.tokenize(current_permissions[new_resource_type].search)
+      unless new_search == current_permissions[new_resource_type].search
+        #new_search can be 0: '' 1: 'new_condition' 2:'current_condition and (new_condition)'
+        current_condition = current_permissions[new_resource_type].search
+        if new_search.empty?
+          # just copy my search [0]
+          self.search = current_condition
+          # check if the new condition does not begin with old one (not augmented)
+        elsif new_search_tokens.take(current_search_tokens.size) != current_search_tokens
+          current_condition << " and ("
+          self.search = "#{current_condition}#{new_search})"
```

Comment:
I don't think we should change the data like this. I understand that you're trying to make it easier for the user but I think we should not save the filter and display an error in this case ideally suggesting what conditions must be present. The reason is that user should know what's the resulting filter.